flopy/mf6/utils/codegen/filters.py

```diff
@@ -0,0 +1,568 @@
+
+                def _should_build(var: dict) -> bool:
+                    subpkg = var.get("ref", None)
+                    if subpkg and component_name != (None, "nam"):
+                        return False
+                    return var["name"] not in [
+                        "simulation",
+                        "model",
+                        "package",
+                        "parent_model",
+                        "parent_package",
+                        "parent_model_or_package",
+                        "parent_file",
+                        "modelname",
+                        "model_nam_file",
+                        "export_netcdf",
+                        "nc_filerecord",
+                        "method",
+                        "interpolation_method_single",
+                        "sfac",
+                        "output",
+                    ]
+
+                stmts = []
+                refs = {}
+                for var in component_vars.values():
+                    name = var["name"]
+                    if name in kwlist:
+                        name = f"{name}_"
+
+                    subpkg = var.get("ref", None)
+                    if _should_build(var):
+                        if subpkg and component_name == (None, "nam"):
+                            stmts.append(
+                                f"self.{'_' if subpkg else ''}{subpkg['key']} "
+                                f"= self.build_mfdata('{subpkg['key']}', None)"
+                            )
+                        else:
+                            _name = (
+                                name[:-1] if name.endswith("_") else name
+                            )
+                            name = name.replace("-", "_")
+                            stmts.append(
+                                f"self.{'_' if subpkg else ''}{name} "
+                                f"= self.build_mfdata('{_name}', {name})"
+                            )
+
+                    if (
+                        subpkg
+                        and subpkg["key"] not in refs
+                        and component_name[1] != "nam"
+                    ):
+                        refs[subpkg["key"]] = subpkg
+                        stmts.append(
+                            f"self._{subpkg['key']} "
+                            f"= self.build_mfdata('{subpkg['key']}', None)"
+                        )
+                        args = (
+                            f"'{subpkg['abbr']}', {subpkg['val']}, "
+                            f"'{subpkg['param']}', self._{subpkg['key']}"
+                        )
+                        stmts.append(
+                            f"self._{subpkg['abbr']}_package "
+                            f"= self.build_child_package({args})"
+                        )
+
+            return stmts
+
+        return list(filter(None, _statements()))
+
+    def safe_name(v: str) -> str:
```

Comment:
These last few simple ones will probably stick around in the long term. The huge mess above, which generates class attributes and init method logic, will go.